src/metpy/calc/indices.py

```diff
@@ -148,6 +152,69 @@ def mean_pressure_weighted(pressure, *args, height=None, bottom=None, depth=None
     return [np.trapz(var_prof * pres_prof, x=pres_prof) / pres_int for var_prof in others]
 
 
+@exporter.export
+@preprocess_and_wrap()
+@check_units('[pressure]')
+def weighted_continuous_average(pressure, *args, height=None, bottom=None, depth=None):
+    r"""Calculate weighted-continuous mean of an arbitrary variable through a layer.
+
+    Layer top and bottom specified in height or pressure.
+
+    Formula based on that from [Holton2004]_ pg. 76 and the NCL function _wgt_vertical_n_
+
+    .. math::  WCA = \frac{\int_{p_s}^{p} A dp}{\int_{p_s}^{p} dp},
+
+    * :math:`WCA` is the weighted continuous average of a variable.
+    * :math:`p` is the bottom pressure level.
+    * :math:`p_s` is the top pressure level.
+    * :math:`A` is the variable whose weighted continuous average is being calculated.
+
+    Parameters
+    ----------
+    pressure : `pint.Quantity`
+        Atmospheric pressure profile.
+
+    args : `pint.Quantity`
+        Parameters for which the weighted-continuous mean is to be calculated.
+
+    height : `pint.Quantity`, optional
+        Heights from sounding. Standard atmosphere heights assumed (if needed)
+        if no heights are given.
+
+    bottom: `pint.Quantity`, optional
+        The bottom of the layer in either the provided height coordinate
+        or in pressure. Don't provide in meters AGL unless the provided
+        height coordinate is meters AGL. Default is the first observation,
+        assumed to be the surface.
+
+    depth: `pint.Quantity`, optional
+        Depth of the layer in meters or hPa.
+
+    Returns
+    -------
+    list of `pint.Quantity`
+        list of layer mean value for each profile in args.
+
+    Notes
+    -----
+    Only functions on 1D profiles (not higher-dimension vertical cross sections or grids).
+    Since this function returns scalar values when given a profile, this will return Pint
+    Quantities even when given xarray DataArray profiles.
+
+    """
+    # Split pressure profile from other variables to average
+    pres_prof, *others = get_layer(
+        pressure, *args, height=height, bottom=bottom, depth=depth
+    )
+
+    # Taking the integral of the weights (pressure) to feed into the weighting
+    # function. Said integral works out to this function:
+    dp = np.gradient(pres_prof)
+
+    # Perform integration on the profile for each variable
+    return [np.sum(dp * var_prof) / np.sum(dp) for var_prof in others]
```

Comment:
Wouldn't this work to use `trapz`?
```suggestion
    # Perform integration on the profile for each variable
    return [np.trapz(var_prof, x=pres_prof) for var_prof in others]
```

I guess the real way to check that is make the code change and see if the tests still pass. 😉

Reply:
>>> import pytest
>>> import xarray as xr
>>>
>>> from metpy.calc import (bulk_shear, bunkers_storm_motion, critical_angle,
...                         mean_pressure_weighted, precipitable_water, significant_tornado,
...                         supercell_composite, weighted_continuous_average)
>>> from metpy.testing import assert_almost_equal, assert_array_almost_equal, get_upper_air_data>>> from metpy.units import concatenate, units

>>> data = get_upper_air_data(datetime(2016, 5, 22, 0), 'DDC')

>>> t, = weighted_continuous_average(data['pressure'],
...                                      data['temperature'],
...                                      height=data['height'],
...                                      depth=6000 * units('meter'))

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/mnt/c/Users/CoeFamily/Documents/GitHub/MetPy/src/metpy/xarray.py", line 1229, in wrapper
    result = func(*bound_args.args, **bound_args.kwargs)
  File "/mnt/c/Users/CoeFamily/Documents/GitHub/MetPy/src/metpy/units.py", line 246, in wrapper
    return func(*args, **kwargs)
  File "/mnt/c/Users/CoeFamily/Documents/GitHub/MetPy/src/metpy/calc/indices.py", line 218, in weighted_continuous_average
    return [np.trapz(var_prof, x=pres_prof) / dp for var_prof in others]
  File "/mnt/c/Users/CoeFamily/Documents/GitHub/MetPy/src/metpy/calc/indices.py", line 218, in <listcomp>
    return [np.trapz(var_prof, x=pres_prof) / dp for var_prof in others]
  File "/home/mariofire/anaconda3/envs/devel/lib/python3.9/site-packages/pint/quantity.py", line 1269, in __truediv__
    return self._mul_div(other, operator.truediv)
  File "/home/mariofire/anaconda3/envs/devel/lib/python3.9/site-packages/pint/quantity.py", line 115, in wrapped
    return f(self, *args, **kwargs)
  File "/home/mariofire/anaconda3/envs/devel/lib/python3.9/site-packages/pint/quantity.py", line 95, in wrapped
    result = f(self, *args, **kwargs)
  File "/home/mariofire/anaconda3/envs/devel/lib/python3.9/site-packages/pint/quantity.py", line 1226, in _mul_div
    raise OffsetUnitCalculusError(self._units, other._units)
pint.errors.OffsetUnitCalculusError: Ambiguous operation with offset unit (degree_Celsius * hectopascal, hectopascal). See https://pint.readthedocs.io/en/latest/nonmult.html for guidance.


When I swap it over to np.trapz, we still need to divide the whole of the function by the sum of dp or it won't be a weighted average. However, when I keep that piece from the previous code, I get this error. I understand it has something to do with the units of the arrays. When run through np.trapz, it returns units of deg_Celsius*hectopascal, which we need to divide by a hectopascal value to remove, which happens with dp. However, it spits out an offset error, which looking at the pint docs points to an error for not declaring the offset I think?

Any help with this is appreciated! If this can't be fixed, then reverting to the method without np.trapz still works.

Reply:
So that error there is due to a bug in `trapz` in Pint (hgrecco/pint#1593).

The correct `trapz` implementation is:
```python
    return [np.trapz(var_prof, x=pres_prof) / (pres_prof[-1] - pres_prof[0])
            for var_prof in others]
```

I'm inclined to go with that since it's simpler and just make sure it gets fixed in `pint`.